File: frontend/src/utils/distance.js
// Radius of Earth in meters
const EARTH_RADIUS = 6371000;

// Convert Degrees → Radians
function toRadians(degree) {
    return degree * (Math.PI / 180);
}

export function calculateDistance(pointA, pointB) {

    const [lat1, lon1] = pointA;
    const [lat2, lon2] = pointB;

    // Convert degrees to radians
    const phi1 = toRadians(lat1);
    const phi2 = toRadians(lat2);

    // Differences
    const deltaPhi = toRadians(lat2 - lat1);
    const deltaLambda = toRadians(lon2 - lon1);

    // Haversine Formula
    const a =
        Math.sin(deltaPhi / 2) * Math.sin(deltaPhi / 2) +
        Math.cos(phi1) *
        Math.cos(phi2) *
        Math.sin(deltaLambda / 2) *
        Math.sin(deltaLambda / 2);

    const c = 2 * Math.atan2(
        Math.sqrt(a),
        Math.sqrt(1 - a)
    );

    return EARTH_RADIUS * c;
}

File: shared/utils/distance.py
from math import atan2, cos, radians, sin, sqrt


EARTH_RADIUS_METERS = 6371000


def calculate_distance(point_a: tuple[float, float], point_b: tuple[float, float]) -> float:
    lat1, lon1 = point_a
    lat2, lon2 = point_b

    phi1 = radians(lat1)
    phi2 = radians(lat2)
    delta_phi = radians(lat2 - lat1)
    delta_lambda = radians(lon2 - lon1)

    haversine_value = (
        sin(delta_phi / 2) ** 2
        + cos(phi1)
        * cos(phi2)
        * sin(delta_lambda / 2) ** 2
    )

    central_angle = 2 * atan2(sqrt(haversine_value), sqrt(1 - haversine_value))
    return EARTH_RADIUS_METERS * central_angle
